task_processing/interfaces/task_executor.py

```diff
@@ -24,14 +21,14 @@ class TaskExecutor(object):
     """
 
     @abc.abstractmethod
-    def run(self, task_config):
-        """Run the supplied task
+    def run(self, task_config, task_id):
```

Comment:
the problem with this solution is that

- we're mutating supposedly immutable task_config (other layers can depend on this)

- retrying executor itself depends on not mutating task_config and there's already a bug in your patch where after second retry on line 65 you'll "restore" not the original task_id but one from the first retry, that you store by mutating task_config in .run() (you restore event.task_id but not event.task_config.task_id)

- we need to maintain id mappings which is lost upon crashing so it makes recovery that much harder (all in-flight tasks must be killed unless original task_id and number of attempts is stored somewhere along them)

- task id mutation is something many layers may choose to implement and it's better to have it done one way than to have to keep track of multiple implementations when debugging, introspecting or implementing layers that have to deal with id components

all of these are addressed by separating task_id out and defining a clear way to add/remove components to it. yes, it increases complexity of everything by a little bit but removes the spikes in complexity of having to deal with all these problems in individual layers.